File: sudoku.c
#include <stdio.h>
#include <time.h>

void sdk(int board[9][9]);
int ck(int board[9][9], int row, int col, int num);
int fill(int board[9][9], int row, int col);
void sdk_print(int board[9][9]);
int main(){
  int board[9][9];
  sdk(board);
  clock_t start = clock();
  int flag = fill(board,0,0);
  if (flag)
    sdk_print(board);
  else{
    printf("no solution!\n");
    sdk_print(board);
  }
  printf("time spend>>%g sec.\n",(clock()-start)/(double)CLOCKS_PER_SEC);
  return 0;
}

void sdk(int board[9][9]){
  for (int i=0;i<9;i++){
    printf("Enter %dth row(9)>>",i+1);
    for (int k=0;k<9;k++)
      scanf("%1d",&board[i][k]);
  }
}

void sdk_print(int board[9][9]){
  for (int i=0;i<9;i++){
    for (int j=0;j<9;j++){
      printf("|%d",board[i][j]);
    }
    printf("|\n");
  }
}

int ck(int board[9][9], int row, int col, int num){
  for (int i=0;i<9;i++){
    if (board[row][i]==num || board[i][col]==num)
      return 0;
  }
  for (int i=0;i<3;i++){
    for (int j=0;j<3;j++){
      if (board[(row/3)*3+i][(col/3)*3+j] == num)
	return 0;
    }
  }
  return 1;
}

int fill(int board[9][9], int row, int col){
  if (col == 9){
    row++;
    col=0;
    if (row == 9)
      return 1;
  }
  if (board[row][col]!=0)
    return fill(board,row,col+1);
  else
    for (int i=1;i<=9;i++){
      if (ck(board,row,col,i)){
	board[row][col]=i;
	if (fill(board,row,col+1))
	  return 1;
	else
	  board[row][col]=0;
      }
    }
  return 0;
}

File: sudoku.py
from time import time
class Solution:
    def solveSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        def check(x, y, s):
            for i in range(9):
                if board[i][y] == s or board[x][i] == s:
                    return False
            for i in [0, 1, 2]:
                for j in [0, 1, 2]:
                    if board[x//3*3+i][y//3*3+j] == s:
                        return False
            return True
        
        def bt(cur):
            if cur == 81:
                return True
            x, y = cur // 9, cur % 9
            if board[x][y] != '.':
                return bt(cur + 1)
            for i in range(1, 10):
                s = str(i)
                if check(x, y, s):
                    board[x][y] = s
                    if bt(cur + 1):
                        return True
                    board[x][y] = '.'
            return False
        
        bt(0)
board=[[".","9","4",".",".","5",".",".",".",],["3",".",".",".",".","9",".","6",".",],["1",".",".",".","8",".","9",".",".",],[".",".",".","7",".","6",".","1","2",],[".",".","3",".",".",".","4",".",".",],["2","1",".","9",".","8",".",".",".",],[".",".","1",".","7",".",".",".","3",],[".","4",".","8",".",".",".",".","9",],[".",".",".","5",".",".","6","2",".",]]
start = time()
Solution().solveSudoku(board)
for i in range(9):
    print(board[i])
    print("\n")
print(time()-start)
